spacepy/toolbox/__init__.py

```diff
@@ -1536,13 +1560,22 @@ def windowMean(data, time=[], winsize=0, overlap=0, st_time=None, op=np.mean):
     op : callable (optional)
         the operator to be called, default numpy.mean
 
+    Raises
+    ------
+    ValueError
+        data and time have different lengths
+    TypeError
+        winsize or overlap are not datetime.timedelta
+    ValueError
+        The requested overlap is larger than requested window
+
```

Comment:
Window size and overlap can be integers (number of array elements).  You can't mix and match though, and supplying a time array through the keyword argument implies time-based binning. If you're using time-based then `timedelta`s have to be used, if you're using index-based binning then indices need to be passed.

Reply:
I'll poke here. The docstring and the code error messages got confusing on what was required.

Reply:
@drsteve What do I not understand here about the flow? `timedelta` is pretty clearly required. 

```
        try:
            assert type(winsize) == datetime.timedelta
            assert type(overlap) == datetime.timedelta
        except AssertionError:
            raise TypeError('windowmean error: winsize/overlap must be timedeltas')
```

Reply:
Looks like we need a unit test for the non-timedelta case. I wonder something got refactored poorly at some point.

Reply:
@balarsen `timedelta` is very clearly required *if* a time array is supplied.
The section you're pasting from starts with:
```
if len(time) == 0:
        startpt, res = 0, 1
        time = list(range(len(data)))
        pts = True
    else:
```
and then the time-based stuff is in the `else`.
Similarly, if you come down past the error checking then the whole thing is in an `if` control block. 
```
    if pts:
        #loop for fixed number of points in window
        ...
    else:
        #loop with time-based window
```
Probably the thing that's missing is a clear statement that *if* a time array is supplied *then* the function assumes time-based windowing and requires `winsize` and `overlap` to be `timedelta`s. If a time array is not supplied the expectation is points-based windowing and `winsize` and `overlap` are expected to be integer.

Reply:
To me it's not clear that `time` being of non-zero length means `time` must be `datetime` and `winsize`, `overlap` must be `timedelta`, since the docs say `time` can be numeric or `datetime`. If `time` is numeric I wouldn't expect `winsize` to be `timedelta`. Am I missing something?

Reply:
The docs don't line up with the code here, and without doing archaeology I'm not sure whether this was functionality that used to exist, or whether the docs are just misleading allowing time to be numeric. However, if we're making sure the documentation lines up with behaviour then supplying a time array pushes us into checking for `winsize` and `overlap` being `datetime.timedelta`.
```
In [13]: tb.windowMean([3]*20, time=range(20))
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/toolbox/__init__.py in windowMean(data, time, winsize, overlap, st_time, op)
   1601         try:
-> 1602             assert type(winsize) == datetime.timedelta
   1603             assert type(overlap) == datetime.timedelta

AssertionError:

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-13-2f219d46f123> in <module>
----> 1 tb.windowMean([3]*20, time=range(20))

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/toolbox/__init__.py in windowMean(data, time, winsize, overlap, st_time, op)
   1603             assert type(overlap) == datetime.timedelta
   1604         except AssertionError:
-> 1605             raise TypeError('windowmean error: winsize/overlap must be timedeltas')
   1606         pts = False #force time-based averaging
   1607         if (type(time[0]) != datetime.datetime):

TypeError: windowmean error: winsize/overlap must be timedeltas
```
If the time input then isn't `datetime` it'll error at trying to work add timedeltas to numeric values:
```
----> 1 tb.windowMean([3]*20, time=range(20), winsize=dt.timedelta(1), overlap=dt.timedelta(0))

~/miniconda3/envs/default/lib/python3.8/site-packages/spacepy/toolbox/__init__.py in windowMean(data, time, winsize, overlap, st_time, op)
   1641     else:
   1642         #loop with time-based window
-> 1643         lastpt = time[0] + winsize
```

Omitting the time array allows us to use numeric window size and overlap since it assumes the indices as timestamps and then indices, winsize and overlap are all on the same scale:
```
In [15]: tb.windowMean([3]*20, winsize=4, overlap=0)
Out[15]: ([3.0, 3.0, 3.0, 3.0], [2.0, 6.0, 10.0, 14.0])
```

For fixing documentation I think clarifying current behaviour is the right thing to do. Moving forward it's probably good to support non-datetime timestamps.

Reply:
I checked the history. From the first CVS commit, `winsize` and `overlap` had to be `timedelta`. At various points in the history, there were some `num2date` things going on that I didn't look at too closely, but might have been trying to make matplotlib timestamps work.

So I agree, document and maybe open an enhancement issue to support non-`datetime`. `rebin` might actually be useful for that.

Reply:
I got this cleaned to I think an acceptable min state in the error message.

Reply:
good in 1aa9067e02c6d8d26ecd3023fbc349c3cd062df2